Mitigate parallel pull issues

tonistiigi opened the pull request:
Attempting to find a solution for the errors people keep getting during parallel pulls.

The fix itself is probably bit controversial. I'm fine of doing this with another layer of synchronization around image extraction or moving the current lock mechanism outside of `Register` function.

What I think is going on:

1) layer is downloaded
2) download is over, pool is released but layer is not in the graph but waits its parent layers to finish.
3) another pull downloads the same layer
4) layer is added to the graph
5) now the parents of the first download finish. first layer is added to the graph but fails because it is already there.

pre-1.7.1 did not have locks in `graph.Register` so these calls could have happened in very unpredictable way, most likely failing in a driver call. After `1.7.1` there is a layer ID based lock in `graph.Register` but you still get an error if you attempt to run it twice for same ID.

As a bonus, there is a defer check in `graph.Register` that removes the image in case of an error. So when this multiple insert happens the second image actually removes the first one, possibly breaking a pull that was already successful?

_All this is just an assumption from looking at the code and error logs from other people. Haven't been able to reproduce the bug nor validate the fix._

cc @stevvooe 

Signed-off-by: Tonis Tiigi tonistiigi@gmail.com

tonistiigi committed:
Mitigate parallel pull issues

Signed-off-by: Tonis Tiigi <tonistiigi@gmail.com>

tonistiigi commented:
Other way to look at this is that the download return channel is buffered. It probably shouldn't but it has a comment that it fixes some issue.

dmp42 commented:
cc @docker/distribution-maintainers

stevvooe commented on graph/graph.go:
```diff
@@ -235,6 +235,11 @@ func (graph *Graph) Register(img *image.Image, layerData archive.Reader) (err er
 	defer graph.imageMutex.Unlock(img.ID)
 
+	// Skip register if image is already registered
+	if graph.Exists(img.ID) {
```
This changes the contract of the method, somewhat. Before, one could get an error and know that they need to close the `archive.Reader`. Now, they have now way of knowing that is has been consumed.

Either way, this method *should* called `layerData.Close` to maintain the contract of the `archive.Reader`, which is really just a `io.ReadCloser`.

stevvooe replied:
Note, the above is just a consideration, rather than a rejection.

stevvooe commented on graph/graph.go:
```diff
@@ -245,11 +250,6 @@ func (graph *Graph) Register(img *image.Image, layerData archive.Reader) (err er
 
-	// (This is a convenience to save time. Race conditions are taken care of by os.Rename)
-	if graph.Exists(img.ID) {
-		return fmt.Errorf("Image %s already exists", img.ID)
```
Interesting. This may the root cause of the pull corruption. Above, we check for `nil` against the error and remove if there is an error. If the image exists, we effectively remove it.

I'd like to see if #9718 gets addressed with this fix.

stevvooe commented on graph/graph.go:
```diff
@@ -517,6 +517,9 @@ func (graph *Graph) saveSize(root string, size int64) error {
 
 // SetDigest sets the digest for the image layer to the provided value.
 func (graph *Graph) SetDigest(id string, dgst digest.Digest) error {
+	graph.imageMutex.Lock(id)
```
If we lock in `SetDigest`, should we not lock in `GetDigest`?

vbatts replied:
... maybe, but it might not matter. And not having it in getdigest could avoid a deadlock.

tonistiigi commented:
I moved the `ReadCloser` change to #15735 to keep this on topic. Also added locking to `graph.GetDigest`.

Please note that this is far from perfect. Layers can be still downloaded twice when there is no actual need for it. But it may be small enough change that we feel comfortable with adding it to `v1.8.2`.

stevvooe commented:
@tonistiigi LGTM.

Let's take an action item to see if can test for a fix to https://github.com/docker/docker/issues/9718.

stevvooe commented:
@vbatts PTAL

rhatdan commented:
@rhvgoyal PTAL

rhatdan commented:
I have a report of this problem in 

https://bugzilla.redhat.com/show_bug.cgi?id=1252421

They have reported that this pull request fixes the problem for them.

spinolacastro commented:
I can confirm the issue was gone on docker-1.7.1 + devicemapper driver applying the following patch:

```
diff --git a/graph/graph.go b/graph/graph.go
index f305622..28ab1f4 100644
--- a/graph/graph.go
+++ b/graph/graph.go
@@ -154,6 +154,10 @@ func (graph *Graph) Register(img *image.Image, layerData archive.ArchiveReader)
        // this doesn't mean Register is fully safe yet.
        graph.imageMutex.Lock(img.ID)
        defer graph.imageMutex.Unlock(img.ID)
+
+        if graph.Exists(img.ID) {
+               return nil
+       }

        defer func() {
                // If any error occurs, remove the new dir from the driver.
@@ -164,11 +168,6 @@ func (graph *Graph) Register(img *image.Image, layerData archive.ArchiveReader)
                }
        }()

-       // (This is a convenience to save time. Race conditions are taken care of by os.Rename)
-       if graph.Exists(img.ID) {
-               return fmt.Errorf("Image %s already exists", img.ID)
-       }
-
        // Ensure that the image root does not exist on the filesystem
        // when it is not registered in the graph.
        // This is common when you switch from one graph driver to another
```

LK4D4 commented:
@tonistiigi need rebase :/

tonistiigi commented:
@LK4D4 done

stevvooe commented:
@spinolacastro Thank you for the confirmation.

LGTM, once again. Also confident this close #9718.

LK4D4 commented:
LGTM
